##   Model–I: ANN without hyperparameter tuning

### Libraries Used

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


In [12]:
import warnings
from sklearn.exceptions import UndefinedMetricWarning

warnings.filterwarnings("ignore", category=UndefinedMetricWarning)


### Step 1: Load the Dataset

In [2]:
df = pd.read_csv('SpotifyAudioFeaturesApril2019.csv')
df.head()


,artist_name,track_id,track_name,acousticness,danceability,duration_ms,energy,instrumentalness,key,liveness,loudness,mode,speechiness,tempo,time_signature,valence,popularity
0,YG,2RM4jf1Xa9zPgMGRDiht8O,"Big Bank feat. 2 Chainz, Big Sean, Nicki Minaj",0.005820,0.743,238373,0.339,0.000,1,0.0812,-7.678,1,0.4090,203.927,4,0.118,15
1,YG,1tHDG53xJNGsItRA3vfVgs,BAND DRUM (feat. A$AP Rocky),0.024400,0.846,214800,0.557,0.000,8,0.2860,-7.259,1,0.4570,159.009,4,0.371,0
2,R3HAB,6Wosx2euFPMT14UXiWudMy,Radio Silence,0.025000,0.603,138913,0.723,0.000,9,0.0824,-5.890,0,0.0454,114.966,4,0.382,56
3,Chris Cooq,3J2Jpw61sO7l6Hc7qdYV91,Lactose,0.029400,0.800,125381,0.579,0.912,5,0.0994,-12.118,0,0.0701,123.003,4,0.641,0
4,Chris Cooq,2jbYvQCyPgX3CdmAzeVeuS,Same - Original mix,0.000035,0.783,124016,0.792,0.878,7,0.0332,-10.277,1,0.0661,120.047,4,0.928,0


### Step 2: Create Target Variable

In [3]:
# Binary classification based on popularity
df['label'] = df['popularity'].apply(lambda x: 1 if x >= 50 else 0)


### Interpretation:
The popularity score is converted into a binary class to simplify the problem into classification. Songs with popularity ≥ 50 are labeled as 1 (popular), while songs with popularity < 50 are labeled as 0 (not popular). This allows the ANN to learn and predict whether a song is popular based on its audio features.

### Step 3: Select Features and Target

In [4]:
features = [
    'danceability', 'energy', 'valence',
    'tempo', 'loudness', 'acousticness'
]

X = df[features]
y = df['label']


### Interpretation:
The selected audio features represent key musical characteristics of each song and are used as input variables (X) for the ANN, while the binary popularity label (y) serves as the target output that the model learns to predict.

### Step 4: Train–Test Split

In [5]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


### Interpretation:
The dataset is split into training (80%) and testing (20%) sets to evaluate the model’s performance on unseen data. The random_state=42 ensures reproducibility of the split.

### Step 5: Feature Scaling

### ANNs perform better when inputs are normalized.

In [6]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


### Interpretation:
Feature scaling is applied to standardize the input data so that all features have zero mean and unit variance. This improves ANN training stability and ensures that features with larger numerical ranges do not dominate the learning process.

### Step 6: Build ANN Model (Fixed Architecture)

In [9]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Input
from tensorflow.keras.optimizers import Adam

model = Sequential()

model.add(Input(shape=(X_train.shape[1],)))
model.add(Dense(16, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(
    optimizer=Adam(),
    loss='binary_crossentropy',
    metrics=['accuracy']
)


### Interpretation:
An Artificial Neural Network is constructed with an explicit input layer, two hidden layers using ReLU activation to learn non-linear patterns, and a sigmoid output layer for binary classification. The model is compiled using the Adam optimizer and binary cross-entropy loss, which are suitable for predicting whether a song is popular or not.

### Step 7: Train the Model

In [10]:
history = model.fit(
    X_train, y_train,
    epochs=20,
    batch_size=32,
    validation_split=0.2,
    verbose=1
)


Epoch 1/20
2614/2614 ━━━━━━━━━━━━━━━━━━━━ 9s 3ms/step - accuracy: 0.8608 - loss: 0.3791 - val_accuracy: 0.8694 - val_loss: 0.3605
Epoch 2/20
2614/2614 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.8680 - loss: 0.3619 - val_accuracy: 0.8693 - val_loss: 0.3589
Epoch 3/20
2614/2614 ━━━━━━━━━━━━━━━━━━━━ 8s 3ms/step - accuracy: 0.8680 - loss: 0.3605 - val_accuracy: 0.8693 - val_loss: 0.3583
Epoch 4/20
2614/2614 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.8680 - loss: 0.3595 - val_accuracy: 0.8693 - val_loss: 0.3580
Epoch 5/20
2614/2614 ━━━━━━━━━━━━━━━━━━━━ 8s 3ms/step - accuracy: 0.8680 - loss: 0.3591 - val_accuracy: 0.8694 - val_loss: 0.3580
Epoch 6/20
2614/2614 ━━━━━━━━━━━━━━━━━━━━ 8s 3ms/step - accuracy: 0.8680 - loss: 0.3588 - val_accuracy: 0.8694 - val_loss: 0.3576
Epoch 7/20
2614/2614 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.8680 - loss: 0.3584 - val_accuracy: 0.8693 - val_loss: 0.3588
Epoch 8/20
2614/2614 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.8680 - loss: 0.3581 - 

### Interpretation:
The ANN is trained on the scaled training data for 20 epochs using a batch size of 32. A portion of the training data (20%) is used for validation to monitor the model’s learning performance and detect overfitting during training.

### Step 8: Model Evaluation

In [13]:
y_pred = (model.predict(X_test) > 0.5).astype(int)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))


817/817 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step
Accuracy: 0.8707381471702446

Confusion Matrix:
 [[22755     0]
 [ 3378     0]]

Classification Report:
               precision    recall  f1-score   support

           0       0.87      1.00      0.93     22755
           1       0.00      0.00      0.00      3378

    accuracy                           0.87     26133
   macro avg       0.44      0.50      0.47     26133
weighted avg       0.76      0.87      0.81     26133



### Interpretation:
The trained ANN predicts the class labels for the test data using a 0.5 decision threshold. Model performance is evaluated using accuracy, confusion matrix, and classification report, which together measure overall correctness, class-wise prediction behavior, and precision–recall performance of the classifier.

### Interpretation of Results

### ->Accuracy

Indicates the percentage of songs correctly classified as popular or not popular.

Example:
Accuracy = 0.78 → 78% of songs classified correctly.